File: Algo/Algorithm/Questions/Arrays/19.zigzagTraverse/main.js
// O(n) Time | O(n) Space where n is the length of the array
function outOfBounds(row, col, height, width) {
  return (row < 0 || row > height) || (col < 0 || col > width);
}
function zigzagTraverse(array) {
  const result = [];
  let row = 0;
  let col = 0;
  const height = array.length - 1;
  const width = array[0].length - 1;
  let goinDown = true;

  while (!outOfBounds(row, col, height, width)) {
    result.push(array[row][col]);

    if (goinDown) {
      if (col === 0 || row === height) {
        goinDown = false;

        if (row === height) {
          col += 1;
        } else {
          row += 1;
        }
      } else {
        row += 1;
        col -= 1;
      }
    } else if (col === width || row === 0) {
      goinDown = true;

      if (col === width) {
        row += 1;
      } else {
        col += 1;
      }
    } else {
      row -= 1;
      col += 1;
    }
  }

  return result;
}

File: Algo/Algorithm/Questions/Arrays/19.zigzagTraverse/main.cpp
#include <vector>
#include <iostream>
using namespace std;

vector<int> zigzagTraverse(vector<vector<int>> array)
{
  vector<int> zigzag = {};
  int row = 0;
  int col = 0;
  int lenRow = array.size() - 1;
  int lenCol = array[0].size() - 1;
  zigzag.push_back(array[row][col]);
  while (row <= lenRow && col <= lenCol)
  {
    if (row < lenRow)
      row++;
    else
      col++;

    while (row >= 0 && col <= lenCol)
    {
      zigzag.push_back(array[row][col]);
      if (row == 0 || col == lenCol)
        break;
      row--;
      col++;
    }

    if (col < lenCol)
      col++;
    else
      row++;

    while (row <= lenRow && col >= 0)
    {
      zigzag.push_back(array[row][col]);
      if (col == 0 || row == lenRow)
        break;
      row++;
      col--;
    }
  }

  return zigzag;
}

int main()
{
  vector<vector<int>>
      array = {{1, 3, 4, 10},
               {2, 5, 9, 11},
               {6, 8, 12, 15},
               {7, 13, 14, 16}};

  for (int x : zigzagTraverse(array))
    cout << x << " ";
  cout << endl;
  return 0;
}
